File: chat-service/src/main/java/org/example/modules/redis/RedisRepository.java
package org.example.modules.redis;

import com.fasterxml.jackson.core.JsonProcessingException;
import io.lettuce.core.api.async.RedisAsyncCommands;
import lombok.RequiredArgsConstructor;
import org.example.domain.ChatMessage;
import org.springframework.data.redis.core.ReactiveRedisTemplate;
import org.springframework.data.redis.core.RedisOperations;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.SetOperations;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Repository
@RequiredArgsConstructor
public class RedisRepository {

    private final RedisTemplate<String,Object> redisTemplate;
    private final SetOperations<String,Object> setOperations;

    private final RedisAsyncCommands<String, String> redisAsyncCommands;

    private final ReactiveRedisTemplate<String, String> reactiveRedisTemplate;

    public void setTypeSave(String key, String value){
        try{
            setOperations.add(key,value);
        }catch (Exception e){
            e.printStackTrace();
        }
    }

    public Set<Object> findWithSetTypeByAll(String key){
        try{
            return setOperations.members(key);
        } catch (Exception e){
            e.printStackTrace();
        }

        return null;
    }

    public Set<String> findWithSetTypeByAllToString(String key) {
        Set<Object> set = findWithSetTypeByAll(key);
        if (set == null) {
            return Collections.emptySet();
        }

        return set.stream()
                .map(Object::toString)
                .collect(Collectors.toSet());
    }

    public void listTypeSave(String key, String value){
        try{
            redisTemplate.opsForList().rightPush(key,value);
        }catch (Exception e){
            e.printStackTrace();
        }
    }

    public List<Object> findWithListType(String key, long start, long end){
        RedisOperations<String, Object> operations = redisTemplate.opsForList().getOperations();
        try{
            return operations.opsForList().range(key,start,end);
        }catch (Exception e){
            e.printStackTrace();
        }

        return null;
    }

    public List<Object> findWithListTypeByAll(String key){
        try{
            return findWithListType(key,0,-1);
        }catch (Exception e){
            e.printStackTrace();
        }

        return null;
    }

    public boolean existsInSet(String key, String value) {
        return Boolean.TRUE.equals(redisTemplate.opsForSet().isMember(key, value));
    }


    public Mono<Boolean> saveWhitList(String key, String value) {
        return reactiveRedisTemplate.opsForList()
                .rightPush(key, value)
                .map(count -> true);
    }

    public Flux<String> findByKey(String key, long start, long end) {
        return reactiveRedisTemplate.opsForList()
                .range(key, start, end)
                .flatMap(json -> {
                    return Mono.just(json);
                });
    }

    public Mono<Boolean> saveWithHash(String key, Map<String,String> info) {
        return reactiveRedisTemplate.opsForHash()
                .putAll(key, info);
    }

    public Mono<Map<String, String>> findByHash(String key) {
        return reactiveRedisTemplate.opsForHash()
                .entries(key)
                .collectMap(
                        entry -> (String) entry.getKey(),
                        entry -> (String) entry.getValue()
                );
    }

    public Mono<Boolean> deleteHash(String key) {
        return reactiveRedisTemplate.delete(key)
                .map(deletedCount -> deletedCount > 0);
    }

//    public Mono<Long> addToSet(String key, String value) {
//        return reactiveRedisTemplate.opsForSet().add(key, value);
//    }
    public Mono<Long> addToSet(String key, String value) {
        return reactiveRedisTemplate.opsForSet().add(key, value);
    }

    public Flux<String> findBySet(String key){
        return  reactiveRedisTemplate.opsForSet()
                .members(key);
    }

    public Mono<Set<String>> getAllSetMembersAsSet(String key) {
        return reactiveRedisTemplate.opsForSet()
                .members(key) // Flux<String>
                .collect(Collectors.toSet()); // Mono<Set<String>>
    }

    public Mono<Long> removeFromSet(String key, String value) {
        return reactiveRedisTemplate.opsForSet().remove(key, value);
    }
}

File: chat-server-kotiln/src/main/kotlin/com/chat/chatserverkotiln/module/redis/RedisRepository.kt
package com.chat.chatserverkotiln.module.redis

import kotlinx.coroutines.reactor.awaitSingle
import org.springframework.data.redis.core.ReactiveRedisTemplate
import org.springframework.data.redis.core.ReactiveSetOperations
import org.springframework.stereotype.Repository
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Repository
class RedisRepository(
    private val reactiveRedisTemplate: ReactiveRedisTemplate<String, String>
) {
    private val setOps: ReactiveSetOperations<String, String>
        get() = reactiveRedisTemplate.opsForSet()

    suspend fun setTypeSave(key: String, value: String): Long {
//        return setOps.add(key, value)
        return setOps.add(key, value).awaitSingle()
    }
    fun setTypeSaveByWebFlux(key: String, value: String): Mono<Long> {
        return reactiveRedisTemplate.opsForSet().add(key, value)
    }

    fun getAllSetMembers(key: String): Flux<String> {
        return reactiveRedisTemplate.opsForSet().members(key)
    }
    suspend fun findWithSetTypeByAll(key: String): List<String> {
        return setOps.members(key)
            .collectList()
            .awaitSingle()
    }

    suspend fun saveWithHash(key: String, info: Map<String, String>): Boolean {
        return reactiveRedisTemplate.opsForHash<String, String>()
            .putAll(key, info)
            .awaitSingle()
    }

    fun saveWithHashByWebFlux(key: String, info: Map<String, String>): Mono<Boolean> {
        return reactiveRedisTemplate.opsForHash<String, String>()
            .putAll(key, info)
            .then(Mono.just(true))
    }

    fun deleteHashByWebFlux(key:String): Mono<Long>{
        return reactiveRedisTemplate.delete(key);
    }
    suspend fun deleteHash(key: String): Long {
        return reactiveRedisTemplate.delete(key).awaitSingle()
    }
    
    suspend fun removeFromSet(key: String, value: String): Long {
        return setOps.remove(key, value).awaitSingle()
    }

    fun removeSetByWebflux(key: String, value : String) : Mono<Long>{
        return setOps.remove(key,value);
    }
    
    suspend fun countKeys(pattern: String): Long {
        return reactiveRedisTemplate.keys(pattern)
            .count()
            .awaitSingle()
    }

    fun countKeysByWebflux(pattern: String) : Mono<Long>{
        return reactiveRedisTemplate.keys(pattern).count();
    }
    
    fun listTypeSave(key: String, value: String): Mono<Long> {
        return reactiveRedisTemplate.opsForList().rightPush(key, value)
    }

    fun findWithListType(key: String, start: Long, end: Long): Flux<String> {
        return reactiveRedisTemplate.opsForList().range(key, start, end)
    }

    fun findWithListTypeByAll(key: String): Flux<String> {
        return findWithListType(key, 0, -1)
    }

    fun findByHash(key: String, subKey: String): Mono<String> {
        return reactiveRedisTemplate.opsForHash<String, String>().get(key, subKey)
    }

    fun existsInSet(key: String, value: String): Mono<Boolean> {
        return setOps.isMember(key, value)
            .map { it == true } // null 안전 처리
    }
}